src/hardware/serialport/softmodem.cpp

```diff
@@ -29,6 +29,75 @@
 #include "softmodem.h"
 #include "misc_util.h"
 
+static const char phoneValidChars[] = "01234567890*=,;#+>";
+
+class CPhonebookEntry {
+public:
+	CPhonebookEntry(const char *_phone, const char *_address) {
+		safe_strncpy(phone, _phone, sizeof(phone));
+		safe_strncpy(address, _address, sizeof(phone));
+	}
+
+	const char *IsMatchingPhone(const char *input) {
+		if (strcmp(input, phone) != 0)
+			return NULL;
+
+		return address;
+	}
+
+	static char IsValidPhone(const char *input) {
+		for (const char *ch = input; *ch; ch++) {
+			if (!strchr(phoneValidChars, *ch)) {
+				LOG_MSG("SERIAL: Phone %s contains invalid character %c", input, *ch);
+				return false;
+			}
+		}
+
+		return true;
+	}
+
+private:
+	char phone[64];
+	char address[64];
+};
+
+static std::vector<CPhonebookEntry *> phones;
+
+bool MODEM_ReadPhonebook(const char *filename) {
```

Comment:
Suggest trying to leverage as much beneficial C++ as possible in here (and subsequent comments)
`bool MODEM_ReadPhonebook(const std::string& filename)`